hasktorch/src/Torch/Optim/Internal.hs

```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Torch.Optim.Internal where
+
+import Foreign.ForeignPtr
+
+import Torch.Internal.Cast
+import Torch.Internal.Class (Castable(..), CppTuple2(..), CppTuple3(..), CppTuple4(..), CppObject(..))
+import qualified Torch.Internal.Type as ATen
+import qualified Torch.Internal.Managed.Optim as LibTorch
+
+import Torch.Tensor
+import Torch.Autograd
+import Torch.NN
+
+import Data.Default.Class
+
+type OptimizerRef = ForeignPtr ATen.Optimizer
+data OptimizerState option p = OptimizerState option OptimizerRef p
+
+class Optimizer option where
+  initOptimizer :: Parameterized d => option -> d -> IO (OptimizerState option d)
+  step :: Parameterized d => OptimizerState option d -> (d -> IO Tensor) -> IO Tensor
```

Comment:
what is `d` here?

Reply:
`d` should be `model`.
`d` is the same as parameters of model.

The optimizer has model parameters, optimizer state and optimizer option.
`step` function extracts the parameters's pointer from the optimizer(`OptimizerState option d`).
The pointer is not copied. The pointer is also stateful and depends on the optimizer.
I'm using a callback to hide the dependencies of the optimizer and model.

Reply:
Hm, ok, I see now, thanks for clarifying!

I think in `master` we have a problem with parameters being duplicated in memory during optimization. we have to keep a pointer around for the tensor parameters of the current model and, in addition to that, the updated parameters as they are produced by the optimizer step.

libtorch tries to avoid that.

I think we have to rethink the `Parameterized` and the `Optimizer` type classes.
we should have a `runStep` function that takes the current model ADT as input and returns the updated model ADT.

Reply:
This is a good point.

> we should have a runStep function that takes the current model ADT as input and returns the updated model ADT.

If this restriction  is relaxed, I think we can integrate it.
`OptimizerState option model`  of this pr is similar to (modelADT, optimizer) of current `runStep`.